### DISCUSSION TITLE
Startup Services in External Modules

### DISCUSSION THREAD

**Discussion by mikecasas** [oqtane/oqtane.framework#693] [General] [answered]
Is this the way to register services in external modules? I thought by implementing IService, oqtane would pick it up. I don't have it right because I am getting the errror.

```
public class MikeStartup : IService
    {
        public MikeStartup(IServiceCollection services)
        {
            services.AddSingleton<Radzen.DialogService>();
        }
    }
```

![radzen](https://user-images.githubusercontent.com/11261164/90675681-cfbc1300-e228-11ea-8016-b46b32216117.PNG)

**Comment by PizzaConsole** [comment]
for that you should probably do, at least the way I understand it.

```csharp
public class MikeDialogService : Radzen.DialogService, IService
    {
       
    }
```

**Reply by PizzaConsole** [reply]
I did not know about [IServerStartup ](https://github.com/oqtane/oqtane.framework/blob/4d7ec16f3645f44183d95c6b8ea1a67522905c39/Oqtane.Server/Infrastructure/Interfaces/IServerStartup.cs)or [IClientStartup](https://github.com/oqtane/oqtane.framework/blob/598b433cd289929091a173951ada2ceef0a6127e/Oqtane.Shared/Interfaces/IClientStartup.cs)....

**Comment by jimspillane** [comment, marked as answer]
@mikecasas would you want to implement [IServerStartup](https://github.com/oqtane/oqtane.framework/blob/4d7ec16f3645f44183d95c6b8ea1a67522905c39/Oqtane.Server/Infrastructure/Interfaces/IServerStartup.cs) or [IClientStartup](https://github.com/oqtane/oqtane.framework/blob/598b433cd289929091a173951ada2ceef0a6127e/Oqtane.Shared/Interfaces/IClientStartup.cs)?

**Reply by mikecasas** [reply]
Ok, I will check these out. 

Since you responded @jimspillane, is there an example of adding css/js from a nuget package?  I know how to use them when they are url references, but what about references like the ones below?  I just extracted the css/js from the nuget package and used the 'module approach', but I was wonder if there was a better solution.

```
 <script src="_content/Radzen.Blazor/Radzen.Blazor.js"></script>
<link rel="stylesheet" href="_content/Radzen.Blazor/css/default-base.css">
```

**Reply by jimspillane** [reply]
You might want to explore a [Razor class library](https://docs.microsoft.com/en-us/aspnet/core/blazor/components/class-libraries?view=aspnetcore-3.1&tabs=visual-studio) for that functionality.

**Reply by mikecasas** [reply]
I tried IServerStartup and I got the below. Radzen wants a singleton. Thanks, at least I know the framework is picking it up.
```
Error: System.InvalidOperationException: Cannot consume scoped service 'Microsoft.AspNetCore.Components.NavigationManager' from singleton 'Radzen.DialogService'.
```

**Reply by jimspillane** [reply]
Maybe trying like step [6. Use Dialog and Notification components](https://blazor.radzen.com/get-started) will work.  

```
public void ConfigureServices(IServiceCollection services)
    {
        ...
        services.AddScoped<DialogService>();
        services.AddScoped<NotificationService>();
        ...
    }
```

Let us know if you get it working.

**Reply by mikecasas** [reply]
Thanks, yes I implemented it as a scoped service like you mentioned above and it worked. Below is what I used to make it work. BTW, they have some good free controls. I used their calendar control as you can see below. Thanks again.

```
  public class ModuleStartup : IServerStartup
    {
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            
        }

        public void ConfigureMvc(IMvcBuilder mvcBuilder)
        {
           
        }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddScoped<Radzen.DialogService>();
        }
    }
```


![rad-cal](https://user-images.githubusercontent.com/11261164/90688590-ab6a3180-e23c-11ea-8bc0-d36a39577c34.PNG)

**Reply by jimspillane** [reply]
Outstanding!  

I wonder what would happen if there were two Modules/Components on the same page using Radzen.  Would calling `services.AddScoped<Radzen.DialogService>();` twice cause it to blow up?

**Reply by mikecasas** [reply]
Only one way to find out...I'll put that on the todo list.

**Comment by markdav-is** [comment]
@mikecasas were you ever able to find the embedded resources ala _content/Radzen.Blazor/Radzen.Blazor.js or are you using the nuget extraction method?

**Reply by mikecasas** [reply]
@nohorse I never went back and tried to figure it out, so I am just extracting the assets from the nuget package.

**Reply by ADefWebserver** [reply]
I posted a Blog: [Using Radzen In Oqtane Modules](https://blazorhelpwebsite.com/ViewBlogPost/46)

**Reply by mikecasas** [reply]
@ADefWebserver Great job! Maybe there could be an auto-extractor. Maybe a class where you define the assets you want out of the external NuGet package. Just a thought.

**Reply by ADefWebserver** [reply]
> @ADefWebserver Great job! Maybe there could be an auto-extractor. Maybe a class where you define the assets you want out of the external NuGet package. Just a thought.

@mikecasas - Yes to update to the latest Radzen you would want to 'automate' the process. Before that, I REALLY would like to figure out why the _content/ resource cannot be used (sigh).

This thread discusses this issue: https://github.com/oqtane/oqtane.framework/discussions/764

**Comment by mikecasas** [comment]
@sbwalker Would it be possible in an external module to add a nuget package as a resource, and then when the module is uploaded to the framework, the contents of the package would be extracted? I know currently a ```Resource``` only supports a Script or Stylesheet.
